### ISSUE TITLE
additional vec2 operations

### ISSUE THREAD

**Issue by mreinstein** [toji/gl-matrix#259]
@toji  would you be open to a PR which adds these vec2 operations? I've found these are useful in several 2d projects:

```javascript
// returns angle of vec2 expressed in radians
export function getAngle(a) {
  return Math.atan2(a[1], a[0])
}

// set a vec2 based on input angle in radians
export function setAngle(out, angle) {
  const length = vec2.length(out)
  out[0] = Math.cos(angle) * length
  out[1] = Math.sin(angle) * length
  return out
}

export function setLength(out, length) {
  const a = getAngle(vec2)
  out[0] = Math.cos(a) * length
  out[1] = Math.sin(a) * length
  return out
}

export function truncate(out, maxLength) {
  if (vec2.length(out) > maxLength) {
    setLength(out, maxLength)
  }
  return out
}
```

**Comment by stefnotch** [comment]
What's the difference between getLength and [scale](http://glmatrix.net/docs/module-vec2.html)?

**Comment by mreinstein** [comment]
@stefnotch there is no `getLength`

**Comment by stefnotch** [comment]
Oh, my bad, I meant `setLength` and [http://glmatrix.net/docs/module-vec2.html#.scale](http://glmatrix.net/docs/module-vec2.html#.scale)

**Comment by mreinstein** [comment]
> I meant setLength and scale

Ah ok. `setLength(out, 5)` is equivalent to:

```javascript
vec2.normalize(out, out)
vec2.scale(out, out, 5) 
```

> About the getAngle and setAngle..A vector itself doesn't have an angle.

It does have an angle if you are treating a vector as a unit vector, which expresses distance and direction from the origin. How about `getRotation` and `setRotation` ?

**Comment by stefnotch** [comment]
:D Yes, that's right. The name `scale` trips up quite a bunch of people.

I'm aware of that. =3
A more general purpose method would be a `getAngle(vecA, vecB)` which takes 2 vectors and returns the angle. If you want to get the angle from the origin, you would pass `[1, 0]` as the second vector.

Creating a vector from a rotation would probably be pretty useful. An alternative would be a `rotate` function and passing it the vector `[1, 0]`.

@toji vec3 has a bunch of rotate functions, however vec2 doesn't have any. Would a pull request with a vec2 rotate function be appreciated?

**Comment by mreinstein** [comment]
@toji any interest in these additional vec2 ops?

**Comment by mreinstein** [comment]
It's been too long since I last looked at this, which of the 4 functions I initially posted would you like to see in the PR? All of them accepted?

**Comment by stefnotch** [comment]
@mreinstein I'd like to keep everything nice and consistent. As it stands, `vec3` has a [angle function which calculates the angle between 2 vectors](http://glmatrix.net/docs/module-vec3.html#.angle) and a bunch of [rotate functions which rotate a vector](http://glmatrix.net/docs/module-vec3.html#.rotateX). I would like to have those 2 functions for `vec2`s. (It doesn't make sense to have more than 1 rotate function for a 2D vector, right?) 😃 

Your suggested `getAngle(vec)` can then be expressed as: `angle(vec, [0,0])`.
And, your suggested `setAngle(out, angle)` can then be written like this: `rotate(vec2.create(), angle)`.

`truncate` will stay a `feature-request` for now. 

tl;dr;
I would love a PR with
- `vec2.angle(a, b)`
- `vec2.rotate(a, angle)`
- A bunch of unit tests

**Comment by mreinstein** [comment]
regarding the unit testing library, is there any particular reason why this won't work:

```javascript
let out = [0, 0];
vec2.rotate(out, [7, 0], Math.PI);

expect(out).toBeEqualish([-7, 0]);
```

this fails:
```
1 failing

  1) vec2
       rotate
         should modify output argument:
     AssertionError [ERR_ASSERTION]: 14
      at Object.toBeEqualish (spec/helpers/spec-helper.js:77:28)
      at Context.<anonymous> (spec/gl-matrix/vec2-spec.js:619:25)
```

despite the fact that printing `out` to the console shows it is in indeed "equalish": 

```
[ -7, 8.572527594031472e-16 ]
```

I've also tried directly comparing individual components of `out`. This also fails:

```javascript
expect(out[0]).toEqual(-7);
```

what gives?

**Comment by stefnotch** [comment]
That's odd. You could push the code that doesn't work to your fork and I'll take a look at it.

**Comment by stefnotch** [comment]
You apparently have to wrap the code into `beforeEach`s:
```javascript
beforeEach(function() {
            result = vec2.rotate(out, [7, 0], Math.PI/2);
        });
```


I'm not sure if this is related to your issue, but, the unit test library has it's limitations. This code doesn't do what you would expect it to do:
```javascript
vec2.rotate(out, [7, 0], Math.PI);
it("should modify output argument", function() { expect(out).toBeEqualish([-7, 0]); });

vec2.rotate(out, [7, 0], Math.PI * 1.5);
it("should modify output argument", function() { expect(out).toBeEqualish([0, -7]); });

vec2.rotate(out, [7, 0], Math.PI * 2.0);
it("should modify output argument", function() { expect(out).toBeEqualish([7, 0]); });
```

Those things won't be executed in that order. As far as I know, the order they will be executed in is this:
```javascript
vec2.rotate(out, [7, 0], Math.PI);
vec2.rotate(out, [7, 0], Math.PI * 2.0);
vec2.rotate(out, [7, 0], Math.PI * 1.5);
it("should modify output argument", function() { expect(out).toBeEqualish([-7, 0]); });
it("should modify output argument", function() { expect(out).toBeEqualish([0, -7]); });
it("should modify output argument", function() { expect(out).toBeEqualish([7, 0]); });
```

**Comment by stefnotch** [comment]
Oh, I totally messed up the specification of the rotate function. 🤦‍♂️ 

I meant, I would love a 
`vec2.rotate(out, a, b, c)` function (similar to [vec3.rotateX](http://glmatrix.net/docs/module-vec3.html#.rotateX) )

**Comment by mreinstein** [comment]
> Those things won't be executed in that order. As far as I know, the order they will be 
> executed in is this

This unit testing framework feels like it's trying so hard to be helpful that it's doing the opposite. At the risk of starting a flame war, any interest in adopting something simpler and more robust? 

I ❤️ https://github.com/tapjs/node-tap

**Comment by stefnotch** [comment]
I'm open to everything. 😉 https://github.com/toji/gl-matrix/issues/295

**Comment by stefnotch** [comment]
@mreinstein It's because I prefer more than one opinion. 😉 I like to look at things from multiple perspectives, and that forces me to look at more than one or 2 testing frameworks. `node-tap` might be absolutely awesome, but I'll never know how awesome it is, if I have no clue about the other testing frameworks.

> There are a seemingly infinite number of them to evaluate, and realistically you'll never evaluate all of them.

I'm just going to check out the biggest ones that catch my attention.

**Comment by mreinstein** [comment]
> It's because I prefer more than one opinion.   I like to look at things from multiple perspectives,
> and that forces me to look at more than one or 2 testing frameworks. 

Definitely a good mindset!  :) What I'm getting at is there should be specific goals or problems that you're trying to solve by changing the testing framework. Otherwise it's just change for the sake of doing something new, and in that case one might as well stick with what is already in place.

**Comment by stefnotch** [comment]
Oh, yes, you are right. To be honest, I don't actually have a goal that I'm trying to solve by changing the testing framework.

So, I'll keep in mind that there are other, amazing testing frameworks and switch to one of them when I actually do have a goal.

**Comment by mreinstein** [comment]
>  I'll keep in mind that there are other, amazing testing frameworks
> and switch to one of them when I actually do have a goal.

Fair enough. 👍

regarding the original intent of this github issue (additional helpful vec2 ops) I think it might be better if you implement them yourself. It's been too long since I've worked on this, these aren't long functions, and you have very specific things you're looking for in terms of conforming with the vec3 modules. In the time it's taken us to discuss it you probably could have written it 20x over. :)

**Comment by stefnotch** [comment]
@mreinstein 
So, to sum it up:
- One person pointed out that there are unit testing frameworks that are easier to use.
- I'm going to have to implement 2 functions
- I probably should update [the contributing guidelines](https://github.com/toji/gl-matrix#contributing), so that everybody knows which things they have to look at to get their PRs accepted.
- I learned a few things.

**Comment by mreinstein** [comment]
@stefnotch I think that's a fair summary of the stuff covered in this discussion thread. 

The only thing I'd say is if you're planning to actually implement the aforementioned proposed new vec2 functions, you might want to keep this issue open, since closing it implies it's done, or won't be implemented.

**Comment by mreinstein** [comment]
> I implemented the other ones.

@stefnotch thanks!

> Feel free to delete all functions except for truncate from your top comment

I'll make a new issue. This issue is overloaded with a lot of discussion and other things. These probably should have been individual issues when I first submitted anyway.
